service/common/src/main/java/org/apache/polaris/service/events/listeners/FileBufferPolarisPersistenceEventListener.java

```diff
@@ -0,0 +1,276 @@
+        shardCount = polarisConfigurationStore.getConfiguration(null, FeatureConfiguration.EVENT_BUFFER_NUM_SHARDS);
+        int numThreads = polarisConfigurationStore.getConfiguration(null, FeatureConfiguration.EVENT_BUFFER_NUM_THREADS);
+        threadPool = Executors.newScheduledThreadPool(numThreads);
+        kryo.register(PolarisEvent.class);
+        kryo.register(PolarisEvent.ResourceType.class);
+
+        // Recover and push any orphaned files
+        discoverAndPushOrphanedFiles();
+    }
+
+    @Override
+    void addToBuffer(PolarisEvent polarisEvent, CallContext callCtx) {
+        int shardNum = polarisEvent.hashCode() % shardCount;
+        String realmId = callCtx.getRealmContext().getRealmIdentifier();
+        if (!buffers.containsKey(realmId)) {
+            createBuffersForRealm(realmId);
+        }
+        BufferShard bufferShard = buffers.get(realmId).get(shardNum);
+        if (bufferShard == null) {
+            LOGGER.error("No buffer shard found for realm: #{}, shard #{}. Event dropped: {}", realmId, shardNum, polarisEvent);
+            return;
+        }
+
+        kryo.writeObject(bufferShard.output, polarisEvent);
+        bufferShard.output.flush();
+    }
+
+    private void createBuffersForRealm(String realmId) {
+        HashMap<Integer, BufferShard> bufferShardsForRealm = new HashMap<>();
+        buffers.put(realmId, bufferShardsForRealm);
+        for (int i = 0; i < shardCount; i++) {
+            bufferShardsForRealm.put(i, createShard(realmId, i));
+        }
+        ScheduledFuture<?> future = threadPool.schedule(new BufferListingTask(shardCount, realmId), timeToFlush, TimeUnit.MILLISECONDS);
+    }
+
+    private BufferShard createShard(String realmId, int shardNum) {
+        File file = new File(getBufferDirectory(realmId, shardNum) + "buffer_timestamp-" + System.currentTimeMillis());
+        File parent = file.getParentFile();
+        if (parent != null && !parent.exists()) {
+            parent.mkdirs();  // Creates all missing parent directories
+        }
+        try {
+            file.createNewFile();
+            Output output = new Output(new FileOutputStream(file));
+            return new BufferShard(null, file, output);
+        } catch (IOException e) {
+            LOGGER.error("Buffer shard {} was unable to initialize: {}", shardNum, e);
+        }
+        return null;
+    }
+
+    private String getBufferDirectory(String realmId, int shardNum) {
+        return System.getProperty("java.io.tmpdir") + "/" + realmId + "/polaris-event-buffer-shard-" + shardNum + "/";
+    }
+
+    private RealmContext getRealmContext(String realmId) {
+        return () -> realmId;
+    }
+
+    private void discoverAndPushOrphanedFiles() {
+        LOGGER.info("Attempting to discover orphaned event buffer files");
+        File topLevelDirectory = new File(System.getProperty("java.io.tmpdir"));
+        File[] realmFiles = topLevelDirectory.listFiles();
```

Comment:
This isn't recursive; don't we need to list each shard dir within the top level dir?

Reply:
I've written it iteratively; first for-loop on L152 goes through the realm directories, 1st nested for-loop (L156) goes through the buffer shards, 2nd nested for-loop (L160) goes through all the files within the shard.

All realm directories have the same parent directory.

Reply:
(this duplicates my other comment about a single listing operation to cover everything, sorry)